tests/test_framework/utils.py

```diff
@@ -46,6 +56,10 @@ def wait_for(success, timeout=TIMEOUT, debug_fn=None):
             interval = 5
     if time.time() > start_time + timeout:
         raise ValueError("Error waiting for {}", success)
+    if not condition():
```

Comment:
I think it's a bit fragile: it's possible for a caller to provide a `condition` function such as it returns `True` above and `False` here during the second call. It would be more robust to change the loop above to be:
```python
while True:
    if success() or time >= start_time + timeout:
        break
    if condition():
        raise
```

Reply:
That's true. Perhaps we could also do:
```python
while True:
    if time.time() >= start_time + timeout:
        raise
    if not condition():
        raise
    if success():
        return
```

Reply:
I've now updated this function.